lib/common/util.js

```diff
@@ -452,102 +451,228 @@ function shouldRetry(err) {
   return false;
 }
 
-module.exports.shouldRetryErr = shouldRetry;
+util.shouldRetryRequest = shouldRetryRequest;
 
-function makeAuthorizedRequest(config) {
-  var GAE_OR_GCE = !config || (!config.credentials && !config.keyFile);
-  var MAX_RETRIES = config && config.maxRetries || 3;
-  var autoRetry = !config || config.autoRetry !== false ? true : false;
-  var attemptedRetries = 0;
-
-  var missingCredentialsError = new Error();
-  missingCredentialsError.message = [
-    'A connection to gcloud must be established via either a `keyFilename` ',
-    'property or a `credentials` object.',
-    '\n\n',
-    'See the "Getting Started with gcloud" section for more information:',
-    '\n\n',
-    '\thttps://googlecloudplatform.github.io/gcloud-node/#/docs/',
-    '\n'
-  ].join('');
-
-  var authorize;
-
-  if (config.customEndpoint) {
-    // Using a custom API override. Do not use `google-service-account` for
-    // authentication. (ex: connecting to a local Datastore server)
-    authorize = function(reqOpts, callback) {
-      callback(null, reqOpts);
-    };
+/**
+ * Create an Auth Client from Google Auth Library, used to get an access token
+ * for authenticating API requests.
+ *
+ * @param {object} config - Configuration object.
+ * @param {object=} config.credentials - Credentials object.
+ * @param {string=} config.email - Account email address, required for PEM/P12
+ *     usage.
+ * @param {string=} config.keyFile - Path to a .json, .pem, or .p12 keyfile.
+ * @param {array} config.scopes - Array of scopes required for the API.
+ * @param {function} callback - The callback function.
+ */
+function getAuthClient(config, callback) {
+  var googleAuth = new GoogleAuth();
+
+  if (config.keyFile) {
+    var authClient = new googleAuth.JWT();
+    authClient.keyFile = config.keyFile;
+    authClient.email = config.email;
+    authClient.scopes = config.scopes;
+    addScope(null, authClient);
+  } else if (config.credentials) {
+    googleAuth.fromJSON(config.credentials, addScope);
   } else {
-    authorize = gsa(config);
+    googleAuth.getApplicationDefault(addScope);
   }
 
-  function makeRequest(reqOpts, callback) {
-    var tokenRefreshAttempts = 0;
-    reqOpts.headers = reqOpts.headers || {};
+  function addScope(err, authClient) {
+    if (err) {
+      callback(err);
+      return;
+    }
 
-    if (reqOpts.headers['User-Agent']) {
-      reqOpts.headers['User-Agent'] += '; ' + USER_AGENT;
-    } else {
-      reqOpts.headers['User-Agent'] = USER_AGENT;
+    if (authClient.createScopedRequired && authClient.createScopedRequired()) {
+      authClient = authClient.createScoped(config.scopes);
     }
 
-    function onAuthorizedRequest(err, authorizedReqOpts) {
+    callback(null, authClient);
```

Comment:
The function is only defined in the context of `getAuthClient`, so it can't be called from outside. It's just a way to handle any of 3 responses from googleAuth the same way. It seemed overkill to extract it into its own exported function, but if it's too weird, I can do that.